src/dotnet-interactive/Connection/StdIoKernelConnector.cs

```diff
@@ -13,40 +14,73 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+
 using Microsoft.DotNet.Interactive.Commands;
 using Microsoft.DotNet.Interactive.Connection;
 using Microsoft.DotNet.Interactive.Events;
 using Microsoft.DotNet.Interactive.Utility;
+
+using Pocket;
+
+using static Pocket.Logger<Microsoft.DotNet.Interactive.App.Connection.StdIoKernelConnector>;
+
 using CompositeDisposable = Pocket.CompositeDisposable;
 
 namespace Microsoft.DotNet.Interactive.App.Connection;
 
-public class StdIoKernelConnector : IKernelConnector, IDisposable
+public class StdIoKernelConnector : IKernelConnector
 {
     private KernelCommandAndEventReceiver? _receiver;
     private KernelCommandAndEventSender? _sender;
     private Process? _process;
+    private ProxyKernel? _rootProxyKernel;
+
     private RefCountDisposable? _refCountDisposable = null;
 
-    public StdIoKernelConnector(string[] command, Uri kernelHostUri, DirectoryInfo? workingDirectory = null)
+    public string[] Command { get; }
+
+    public string RootProxyKernelLocalName { get; }
+
+    public Uri KernelHostUri { get; }
+
+    public DirectoryInfo WorkingDirectory { get; }
+
+    public int? ProcessId => _process?.Id;
+
+    public StdIoKernelConnector(
+        string[] command,
+        string rootProxyKernelLocalName,
+        Uri kernelHostUri,
+        DirectoryInfo? workingDirectory = null)
     {
         Command = command;
+        RootProxyKernelLocalName = rootProxyKernelLocalName;
         KernelHostUri = kernelHostUri;
         WorkingDirectory = workingDirectory ?? new DirectoryInfo(Environment.CurrentDirectory);
     }
 
-    public string[] Command { get; }
-
-    public DirectoryInfo WorkingDirectory { get; }
-
-    public Uri KernelHostUri { get; }
+    /// <remarks>
+    /// TODO: Does it even make sense to implement <see cref="IKernelConnector"/> here considering that we have
+    /// concepts (such as a single root <see cref="ProxyKernel"/>, and zero or more (child) <see cref="ProxyKernel"/>s)
+    /// that the <see cref="IKernelConnector"/> abstraction does not understand / support? Should the '#!connect stdio'
+    /// command be removed?
+    /// 
+    /// The current implementation only supports creating / retrieving the root <see cref="ProxyKernel"/>.
+    /// </remarks>
+    async Task<Kernel> IKernelConnector.CreateKernelAsync(string kernelName)
+        => await CreateRootProxyKernelAsync();
 
-    public async Task<Kernel> CreateKernelAsync(string kernelName)
+    public async Task<ProxyKernel> CreateRootProxyKernelAsync()
     {
-        ProxyKernel? proxyKernel;
-
-        if (_receiver is null)
+        if (_rootProxyKernel is null)
         {
+            using var activity = Log.OnEnterAndExit();
+
+            Debug.Assert(_receiver is null);
```

Comment:
We don't currently use `Debug.Assert` in this codebase. Let's have a discussion before deciding to introduce it.

Reply:
I've gone ahead and removed them for now.